the-collab-lab/tcl-43-smart-shopping-list
# Rt ng estimate next purchase date

## Description

This PR added `dateItemAdded`, `estimatedPurchaseInterval` and `totalPurchases` as properties to items added to the list via the AddItem component. Functionality in the List component was also updated to set the `lastPurchaseDate` to now, increment `totalPurchases` and calculate `estimatedPurchaseInterval` when an item is marked as purchased.  

The `estimatedPurchaseInterval` is a number that represents the number of days until the next purchase date for an item.

The `calculateEstimate` script is from the `@the-collab-lab/shopping-list-utils` dependency that was previously installed. A breakdown of what that script is doing can be found [here](https://github.com/the-collab-lab/shopping-list-utils/blob/main/src/calculateEstimate/index.ts).

## Related Issue

closes #10

## Acceptance Criteria


## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

No UI changes were made.

## Testing Steps / QA Criteria

* From your terminal, pull down this branch with `git fetch` followed by `git switch rt-ng-estimate-next-purchase-date`
* Run `npm start` to launch the app
* In your browser, go to Application > Storage > and then click on "Clear Site Data" to delete your user token. Refresh the browser page and create a new shopping list and add some new items to that list. _This is important to follow, as any previous lists will not work since the items contained on those lists do not have the necessary properties._
*  Once new items have been added, mark an item or two as purchased on the list view
* Navigate to the Firestore and select the collection associated with your new user token
* For one of the items that you have checked as purchased, edit the `totalPurchases` to a number greater than 2, and subtract 500,000 from the `lastPurchaseDate`
* Go back to the app, and recheck that item to mark it as purchased
* Navigate again to Firestore and the `totalPurchases` property should be incremented by 1 from what you previous edited the number to be, and `estimatedPurchaseInterval` should be a number different from 0.
